gqlalchemy/instance_runner.py

```diff
@@ -0,0 +1,287 @@
+# Copyright (c) 2016-2022 Memgraph Ltd. [https://memgraph.com]
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import docker
+import os
+import psutil
+import socket
+import subprocess
+import time
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import Any, Dict, Union
+from .memgraph import Memgraph
+
+
+MEMGRAPH_DEFAULT_BINARY_PATH = "/usr/lib/memgraph/memgraph"
+MEMGRAPH_DEFAULT_PORT = 7687
+MEMGRAPH_CONFIG_BOLT_PORT = "--bolt-port"
+MEMGRAPH_CONFIG_BOLT_ADDRESS = "--bolt-address"
+DOCKER_IMAGE_TAG_LATEST = "latest"
+LOOPBACK_ADDRESS = "127.0.0.1"
+WILDCARD_ADDRESS = "0.0.0.0"
+
+TIMEOUT_ERROR_MESSAGE = "Waited too long for the port {port} on host {host} to start accepting connections."
+DOCKER_TIMEOUT_ERROR_MESSAGE = "Waited too long for the Docker container to start."
+MEMGRAPH_CONNECTION_ERROR_MESSAGE = "The Memgraph process probably died."
+
+
+class DockerImage(Enum):
+    MEMGRAPH = "memgraph/memgraph"
+    MAGE = "memgraph/memgraph-mage"
+
+
+class DockerContainerStatus(Enum):
+    EXITED = "exited"
+    PAUSED = "paused"
+    RESTARTING = "restarting"
+    RUNNING = "running"
+
+
+def wait_for_port(
+    host: str = LOOPBACK_ADDRESS, port: int = MEMGRAPH_DEFAULT_PORT, delay: float = 0.01, timeout: float = 5.0
+) -> None:
+    """Wait for a TCP port to become available.
+
+    Args:
+        host: A string representing the IP address that is being checked.
+        port: A string representing the port that is being checked.
+        delay: A float that defines how long to wait between retries.
+        timeout: A float that defines how long to wait for the port.
+
+    Raises:
+      TimeoutError: Raises an error when the host and port are not accepting
+        connections after the timeout period has passed.
+    """
+    start_time = time.perf_counter()
+    time.sleep(delay)
+    retries = 0
+    while True:
+        try:
+            with socket.create_connection((host, port), timeout=timeout):
+                break
+        except OSError as ex:
+            time.sleep(delay * 2**retries)
+            if time.perf_counter() - start_time >= timeout:
+                raise TimeoutError(TIMEOUT_ERROR_MESSAGE.format(port=port, host=host)) from ex
+
+        retries += 1
```

Comment:
Multiplication by 2 is a much less costly operation than power by 2, especially with floats.
Consider doing something like:

```
backoff = delay
while True:
...
time.sleep(backoff)
backoff *= 2
```